js-packages/@prestojs/routing/src/useUrlQueryState.ts

```diff
@@ -0,0 +1,447 @@
+// @flow
+import { useCallback, useEffect, useRef, useMemo } from 'react';
+import pickBy from 'lodash/pickBy';
+import isEqual from 'lodash/isEqual';
+import qs from 'qs';
+
+type Parse = (value: string, paramName: string) => any;
+type Stringify = (value: any, paramName: string) => string;
+
+type Options = {
+    /**
+     * A function used to parse query values to transform them to a certain type or shape.
+     *
+     * It is passed 2 arguments; the value to transform and the name of the parameter.
+     *
+     * eg. Calling `setUrlState({ count: 1, name: 'test' });` would call `parse`
+     * twice, once with `1, 'count'` and once with `'test', name`.
+     */
+    parse?: Parse;
+    /**
+     * A function used to stringify a value to be stored in the URL. If not specified the value's toString() method will be called. This could be useful if needing to store a complex object in the query string.
+     *
+     * It is passed 2 arguments; the value to stringify and the name of the parameter.
+     */
+    stringify?: Stringify;
+    /**
+     * A value to prefix query param keys with. The returned state object
+     * contains the un-prefixed keys.
+     *
+     * If not specified all query params are sync'd. To make this work nicely
+     * with other usages of the hook on the same page consider setting
+     * `controlledKeys`
+     */
+    prefix?: string;
+    /**
+     * If specified only these keys will be synced to and read from the request
+     * URL. Any other keys will be ignored.
+     */
+    controlledKeys?: Array<string>;
```

Comment:
If `controlledKeys == true` then could default to all the keys in the `params` list above (and throw an exception if this is combined with a wildcard/fallback key)